apps/next-evans/lib/residences.ts

```diff
@@ -0,0 +1,227 @@
+import invariant from 'tiny-invariant';
+import type { friend_residences as FriendResidence } from '@prisma/client';
+
+type Map = 'UK' | 'US' | 'Europe';
+
+interface Position {
+  top: number;
+  left: number;
+  map: Map;
+}
+
+export default function getResidences(
+  residences: Array<Pick<FriendResidence, 'city' | 'region'>>,
+): Array<Pick<FriendResidence, 'city' | 'region'> & Position> {
+  const positions = residences.map(getPosition);
+  const map = deriveMap(positions);
+  return residences.map((residence) => {
+    const position = getPosition(residence);
+    if (position.map !== map) {
+      position.top = -1000;
+      position.left = -1000;
+    }
+    return { ...residence, ...position, map };
+  });
+}
+
+function deriveMap(positions: Position[]): 'UK' | 'US' | 'Europe' {
+  if (positions.length === 0) return `UK`;
+  const dict = positions.reduce((acc, pos) => {
+    acc[pos.map] = (acc[pos.map] || []).concat([pos.map]);
+    return acc;
+  }, {} as Record<'UK' | 'US' | 'Europe', Map[]>);
```

Comment:
hmmm... i think i did this `as` cast, but can you try removing it and add the type annotation as an explicit generic and see if that works? `positions.reduce<Record<blah...>>(...)`